Update define checks to fix Linux compilation

isolin opened the pull request:
The following functions were throwing an error:
_mm256_add_ps, _mm256_pow_ps, _mm256_exp_ps, _mm256_expm1_ps, _mm256_log_ps

Added defined(MSVC) check to fallback to the non-AVX2 alternative.

isolin committed:
Update define checks to fix Linux compilation

The following functions were throwing an error:
_mm256_add_ps, _mm256_pow_ps, _mm256_exp_ps, _mm256_expm1_ps, _mm256_log_ps

Added defined(MSVC) check to fallback to the non-AVX2 alternative.

Smilebags commented:
One question I have here is whether we should be changing the definition of __KERNEL_AVX2__ rather than adding two checks here. Does your system actually support AVX2? If so, I wonder whether there should be a separate flag which represents support for these specific methods?

isolin commented:
I am just running further tests and it looks like this pull request was too early - it's wrong. I shouldn't do any pull requests in the night when I'm already tired.

I reverted all the checks back and instead I added the following `include` right before the namespace declaration.
```cpp
#include <immintrin.h>
```

Now the `float8.h` compiles without any problems on my machine. So it was not an MSVC-specific issue after all. Please let me know of what you think of this proposal and whether I should create a pull request for that. It would be great to test if the new include works with other compilers as well.

I checked once more my CPU flags. Neither AVX nor AVX2 are in the list.

Smilebags commented:
Hi, I can't really give advice on what solution would be ideal as I'm very unfamiliar with C/C++ programming and Blender. @pembem22 might be able to offer some advice, so please create a PR and we can discuss it. Thanks for your work on this.

isolin commented:
I am really puzzled. The last test I did with the `#include <immintrin.h>` after reverting all `defined(MSVC)` shows again the problems with the following methods:
`_mm256_exp_ps`, `_mm256_expm1_ps`, `_mm256_log_ps`, `_mm256_pow_ps` so at least `add` is not on the list anymore. 

So I have no idea what is a valid / the best solution for this.

Maybe this could be interesting to look at? https://github.com/vectorclass/version2

pembem22 commented:
Can you check if this `#if defined(__KERNEL_AVX2__) && defined(_INCLUDED_IMM)` works?

isolin commented:
> Can you check if this `#if defined(__KERNEL_AVX2__) && defined(_INCLUDED_IMM)` works?

Thank you very much! Yes, that works. I will update the pull request with the respective changes.

isolin committed:
Improved the defined checks

According to https://github.com/Smilebags/blender/pull/8#issuecomment-636378421

pembem22 commented:
Thanks for testing. This works on Windows as well. @Smilebags I think you can merge this.

Smilebags commented:
Thanks for the PR, @isolin, Linux users everywhere will be rejoicing 😄

isolin commented:
I am glad I could help :) So it's the `CAT` issue that remains to be solved for Linux compilation.

pembem22 commented:
@isolin Are there still any issues with compiling?
